lib/Marlin/Marlin/src/feature/twibus.cpp

```diff
@@ -26,174 +26,235 @@
 
 #include "twibus.h"
 
-#include <Wire.h>
+FORCE_INLINE char hex_nybble(const uint8_t n) {
+  return (n & 0xF) + ((n & 0xF) < 10 ? '0' : 'A' - 10);
+}
+
+TWIBus twibus;
 
 TWIBus::TWIBus() {
-  #if I2C_SLAVE_ADDRESS == 0
-    Wire.begin();                  // No address joins the BUS as the master
-  #else
-    Wire.begin(I2C_SLAVE_ADDRESS); // Join the bus as a slave
-  #endif
   reset();
 }
 
 void TWIBus::reset() {
   buffer_s = 0;
   buffer[0] = 0x00;
+  read_buffer_available = 0;
+  read_buffer_pos = 0;
+}
+
+bool TWIBus::read_buffer_has_byte() {
+  return read_buffer_pos < read_buffer_available;
 }
 
-void TWIBus::address(const uint8_t adr) {
+uint8_t TWIBus::read_buffer_read_byte() {
+  if (!read_buffer_has_byte()) {
+    return 0;
+  }
+  return read_buffer[read_buffer_pos++];
+}
+
+bool TWIBus::address(const uint8_t adr) {
   if (!WITHIN(adr, 8, 127)) {
     SERIAL_ECHO_MSG("Bad I2C address (8-127)");
+    return false;
+  }
+
+  if (isRestrictedAddress(addr)) {
+    SERIAL_ECHO_MSG("Restricted I2C address.");
+    return false;
   }
 
   addr = adr;
 
-  #if ENABLED(DEBUG_TWIBUS)
-    debug(PSTR("address"), adr);
-  #endif
+  debug(F("address"), adr);
+  return true;
+}
+
+bool TWIBus::isRestrictedAddress(uint8_t addr) {
+  switch (addr) {
+    case 0xa6:
+    case 0xae:
+    case 0xa7:
+    case 0xaf:
+      // EEPROM
+      return true;
+    case 0x22:
+    case 0x23:
+      // USBC
+      return true;
+     case 0x30:
+     case 0x31:
+     case 0x32:
+     case 0x33:
+     case 0x34:
+     case 0x35:
+     case 0x36:
+     case 0x37:
+     case 0x38:
+     case 0x39:
+     case 0x3A:
+     case 0x3B:
+     case 0x3C:
+     case 0x3D:
+     case 0x3E:
+     case 0x3F:
+      // IO Extender
```

Comment:
I'm sorry, some of the addresses I gave you were already bitshifted with read/write bit included, so I had to recalculate them:

```
EEPROM:
    0x53, 0x57
USBC:
    0x22, 0x23
IO Extender:
    0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1D, 0x1E, 0x1F
```

Reply:
Replace all restricted addresses with the ones on this list.